File: julia/AIInAction/src/ch124_kmeans_variants.jl
"""
    Ch124KMeansVariants

K-Means variants and extensions (Chapter 124). Mirrors the Python module
`aiinaction.ch124_kmeans_variants` and the Rust module
`aiinaction::ch124_kmeans_variants`. The shared fixtures in
`test/test_ch124_kmeans_variants.jl` match the Python/Rust suites to keep the
three implementations at parity.

Every routine is deterministic given its inputs; initialization is supplied by
the caller as explicit centroids or indices.
"""
module Ch124KMeansVariants

export lloyd_step, inertia, mini_batch_update, kmedians_centroid, pam_assign_cost,
    rbf_kernel_matrix, kernel_assignment_distances, fuzzy_centroids, fuzzy_memberships,
    bisecting_split

_sqdist(a, b) = sum((a .- b) .^ 2)

function _check_matrix(X::AbstractMatrix, name::AbstractString)
    size(X, 1) == 0 && throw(ArgumentError("$name must contain at least one sample"))
    size(X, 2) == 0 && throw(ArgumentError("$name must have at least one feature"))
    return size(X, 2)
end

"""
    lloyd_step(X, centroids) -> (labels, new_centroids)

One iteration of Lloyd's algorithm. `X` is `n x d`, `centroids` is `k x d`.
Assigns to the nearest centroid (ties to the lowest index, 1-based labels) then
recomputes each centroid as the mean of its members; an empty centroid is left
unchanged.
"""
function lloyd_step(X::AbstractMatrix, centroids::AbstractMatrix)
    d = _check_matrix(X, "X")
    k = _check_matrix(centroids, "centroids")
    k == d || throw(ArgumentError("centroid dimension $k does not match data dimension $d"))
    n = size(X, 1)
    kc = size(centroids, 1)
    labels = Vector{Int}(undef, n)
    for i in 1:n
        best, bestd = 1, Inf
        for j in 1:kc
            dd = _sqdist(view(X, i, :), view(centroids, j, :))
            if dd < bestd
                bestd, best = dd, j
            end
        end
        labels[i] = best
    end
    new_c = copy(Matrix{Float64}(centroids))
    for j in 1:kc
        idx = findall(==(j), labels)
        if !isempty(idx)
            new_c[j, :] = vec(sum(view(X, idx, :), dims=1)) ./ length(idx)
        end
    end
    return labels, new_c
end

"""Within-cluster sum of squared distances to the nearest centroid."""
function inertia(X::AbstractMatrix, centroids::AbstractMatrix)
    d = _check_matrix(X, "X")
    k = _check_matrix(centroids, "centroids")
    k == d || throw(ArgumentError("centroid dimension $k does not match data dimension $d"))
    total = 0.0
    for i in 1:size(X, 1)
        best = Inf
        for j in 1:size(centroids, 1)
            dd = _sqdist(view(X, i, :), view(centroids, j, :))
            dd < best && (best = dd)
        end
        total += best
    end
    return total
end

"""
    mini_batch_update(centroids, counts, batch) -> (new_centroids, new_counts)

One mini-batch K-Means update (Sculley, 2010). Points are processed in order with
per-centroid learning rate `1 / count`.
"""
function mini_batch_update(centroids::AbstractMatrix, counts::AbstractVector, batch::AbstractMatrix)
    k = _check_matrix(centroids, "centroids")
    length(counts) == size(centroids, 1) ||
        throw(ArgumentError("counts length $(length(counts)) does not match number of centroids $(size(centroids,1))"))
    any(c -> c < 0, counts) && throw(ArgumentError("counts must be non-negative"))
    bd = _check_matrix(batch, "batch")
    bd == k || throw(ArgumentError("batch dimension $bd does not match centroid dimension $k"))
    c = copy(Matrix{Float64}(centroids))
    cnt = copy(Vector{Float64}(counts))
    for i in 1:size(batch, 1)
        x = view(batch, i, :)
        best, bestd = 1, Inf
        for j in 1:size(c, 1)
            dd = _sqdist(x, view(c, j, :))
            if dd < bestd
                bestd, best = dd, j
            end
        end
        cnt[best] += 1.0
        eta = 1.0 / cnt[best]
        c[best, :] = (1.0 - eta) .* c[best, :] .+ eta .* x
    end
    return c, cnt
end

"""Coordinatewise (lower) median: the L1-optimal k-medians representative."""
function kmedians_centroid(members::AbstractMatrix)
    d = _check_matrix(members, "members")
    n = size(members, 1)
    idx = (n - 1) ÷ 2 + 1  # lower median (1-based)
    out = Vector{Float64}(undef, d)
    for t in 1:d
        out[t] = sort(collect(view(members, :, t)))[idx]
    end
    return out
end

"""
    pam_assign_cost(distances, medoid_indices) -> (labels, total_cost)

Assign each point to its nearest medoid given an `n x n` dissimilarity matrix.
`medoid_indices` and returned labels are 1-based; labels index into
`medoid_indices`.
"""
function pam_assign_cost(distances::AbstractMatrix, medoid_indices::AbstractVector{<:Integer})
    n = size(distances, 1)
    (n > 0 && size(distances, 2) == n) ||
        throw(ArgumentError("distances must be a square n x n dissimilarity matrix"))
    isempty(medoid_indices) && throw(ArgumentError("medoid_indices must be non-empty"))
    for m in medoid_indices
        (1 <= m <= n) || throw(ArgumentError("medoid index $m out of range [1, $n]"))
    end
    labels = Vector{Int}(undef, n)
    total = 0.0
    for i in 1:n
        bestpos, bestcost = 1, Inf
        for (pos, m) in enumerate(medoid_indices)
            c = distances[i, m]
            if c < bestcost
                bestcost, bestpos = c, pos
            end
        end
        labels[i] = bestpos
        total += bestcost
    end
    return labels, total
end

"""Gaussian (RBF) kernel matrix `K[i,j] = exp(-gamma ||x_i - x_j||^2)`."""
function rbf_kernel_matrix(X::AbstractMatrix, gamma::Real)
    gamma > 0 || throw(ArgumentError("gamma must be positive, got $gamma"))
    _check_matrix(X, "X")
    n = size(X, 1)
    K = Matrix{Float64}(undef, n, n)
    for i in 1:n, j in 1:n
        K[i, j] = exp(-gamma * _sqdist(view(X, i, :), view(X, j, :)))
    end
    return K
end

"""
    kernel_assignment_distances(kernel, labels, n_clusters) -> Matrix

Feature-space squared distances from each point to every cluster mean via the
kernel trick. `labels` are 1-based; an empty cluster yields `Inf`.
"""
function kernel_assignment_distances(kernel::AbstractMatrix, labels::AbstractVector{<:Integer}, n_clusters::Integer)
    n = size(kernel, 1)
    (n > 0 && size(kernel, 2) == n) ||
        throw(ArgumentError("kernel must be a square n x n matrix"))
    length(labels) == n ||
        throw(ArgumentError("labels length $(length(labels)) does not match kernel size $n"))
    n_clusters > 0 || throw(ArgumentError("n_clusters must be positive"))
    for c in labels
        (1 <= c <= n_clusters) || throw(ArgumentError("label $c out of range [1, $n_clusters]"))
    end
    members = [findall(==(j), labels) for j in 1:n_clusters]
    third = [isempty(S) ? Inf : sum(kernel[S, S]) / length(S)^2 for S in members]
    out = Matrix{Float64}(undef, n, n_clusters)
    for i in 1:n, j in 1:n_clusters
        S = members[j]
        if isempty(S)
            out[i, j] = Inf
        else
            cross = sum(kernel[i, l] for l in S)
            out[i, j] = kernel[i, i] - 2.0 * cross / length(S) + third[j]
        end
    end
    return out
end

"""Membership-weighted fuzzy c-means centroids."""
function fuzzy_centroids(X::AbstractMatrix, memberships::AbstractMatrix, m::Real)
    m > 1 || throw(ArgumentError("fuzziness exponent m must be greater than 1, got $m"))
    d = _check_matrix(X, "X")
    size(memberships, 1) == size(X, 1) ||
        throw(ArgumentError("memberships must have shape (n_samples, n_clusters)"))
    k = size(memberships, 2)
    num = zeros(Float64, k, d)
    den = zeros(Float64, k)
    for i in 1:size(X, 1), j in 1:k
        w = memberships[i, j]^m
        den[j] += w
        num[j, :] .+= w .* view(X, i, :)
    end
    for j in 1:k
        den[j] == 0 &&
            throw(ArgumentError("a cluster has zero total membership; cannot form its centroid"))
        num[j, :] ./= den[j]
    end
    return num
end

"""Update fuzzy memberships from distances to all centroids (Bezdek FCM)."""
function fuzzy_memberships(X::AbstractMatrix, centroids::AbstractMatrix, m::Real)
    m > 1 || throw(ArgumentError("fuzziness exponent m must be greater than 1, got $m"))
    d = _check_matrix(X, "X")
    kd = _check_matrix(centroids, "centroids")
    kd == d || throw(ArgumentError("centroid dimension $kd does not match data dimension $d"))
    k = size(centroids, 1)
    p = 2.0 / (m - 1.0)
    out = zeros(Float64, size(X, 1), k)
    for i in 1:size(X, 1)
        dist = [sqrt(_sqdist(view(X, i, :), view(centroids, j, :))) for j in 1:k]
        zeros_idx = findall(==(0.0), dist)
        if !isempty(zeros_idx)
            for j in zeros_idx
                out[i, j] = 1.0 / length(zeros_idx)
            end
            continue
        end
        for j in 1:k
            denom = sum((dist[j] / dist[l])^p for l in 1:k)
            out[i, j] = 1.0 / denom
        end
    end
    return out
end

"""
    bisecting_split(X, init_two_centroids) -> (labels, two_centroids, sse)

Run 2-means to convergence to bisect a cluster. Labels are 1-based in {1, 2}.
"""
function bisecting_split(X::AbstractMatrix, init_two_centroids::AbstractMatrix)
    size(init_two_centroids, 1) == 2 ||
        throw(ArgumentError("bisecting requires exactly 2 initial centroids, got $(size(init_two_centroids,1))"))
    centroids = Matrix{Float64}(init_two_centroids)
    labels = Int[]
    prev = nothing
    for _ in 1:1000
        labels, centroids = lloyd_step(X, centroids)
        if prev !== nothing && prev == labels
            break
        end
        prev = copy(labels)
    end
    return labels, centroids, inertia(X, centroids)
end

end # module Ch124KMeansVariants

File: src/aiinaction/ch124_kmeans_variants.py
"""K-Means variants and extensions (Chapter 124).

Production reference implementations of the centroid-based clustering family that
relaxes the assumptions of Lloyd's algorithm: mini-batch K-Means, k-medoids,
k-medians, kernel K-Means, fuzzy c-means, and bisecting K-Means.

The Python implementation here is the executed reference. The Julia module
(``AIInAction.Ch124KMeansVariants``) and the Rust module
(``aiinaction::ch124_kmeans_variants``) mirror the same public API, and the
cross-language parity tests assert that all three agree on the shared fixtures in
``tests/test_ch124_kmeans_variants.py``.

Every routine here is deterministic given its inputs: initialization is supplied by
the caller as explicit centroid indices or centroid coordinates, so the three
language implementations produce bit-comparable results without sharing a random
number generator. This is what makes the parity tests meaningful.
"""
from __future__ import annotations

import math
from collections.abc import Sequence

import numpy as np

__all__ = [
    "lloyd_step",
    "mini_batch_update",
    "kmedians_centroid",
    "pam_assign_cost",
    "kernel_assignment_distances",
    "fuzzy_memberships",
    "fuzzy_centroids",
    "bisecting_split",
    "inertia",
    "rbf_kernel_matrix",
]

Matrix = Sequence[Sequence[float]]
Vector = Sequence[float]


def _as_matrix(x: Matrix, name: str = "X") -> np.ndarray:
    arr = np.asarray(x, dtype=float)
    if arr.ndim != 2:
        raise ValueError(f"{name} must be a 2-D array of shape (n_samples, n_features), got ndim={arr.ndim}")
    if arr.shape[0] == 0:
        raise ValueError(f"{name} must contain at least one sample")
    if arr.shape[1] == 0:
        raise ValueError(f"{name} must have at least one feature")
    return arr


def _as_centroids(c: Matrix, n_features: int) -> np.ndarray:
    arr = np.asarray(c, dtype=float)
    if arr.ndim != 2:
        raise ValueError(f"centroids must be a 2-D array of shape (k, n_features), got ndim={arr.ndim}")
    if arr.shape[0] == 0:
        raise ValueError("centroids must contain at least one centroid")
    if arr.shape[1] != n_features:
        raise ValueError(
            f"centroid dimension {arr.shape[1]} does not match data dimension {n_features}"
        )
    return arr


def _pairwise_sq_dist(X: np.ndarray, C: np.ndarray) -> np.ndarray:
    """Squared Euclidean distances, shape (n, k)."""
    # ||x - c||^2 = ||x||^2 - 2 x.c + ||c||^2, computed stably.
    diff = X[:, None, :] - C[None, :, :]
    return np.einsum("nkd,nkd->nk", diff, diff)


def lloyd_step(X: Matrix, centroids: Matrix) -> tuple[list[int], list[list[float]]]:
    """One iteration of Lloyd's algorithm: assign then update.

    Assigns each point to its nearest centroid by squared Euclidean distance
    (ties broken toward the lowest index), then recomputes each centroid as the
    mean of its assigned points. A centroid that captures no points is left
    unchanged.

    Returns a tuple ``(labels, new_centroids)`` where ``labels[i]`` is the cluster
    index of point ``i`` and ``new_centroids`` is the updated centroid list.

    >>> labels, c = lloyd_step([[0.0], [0.2], [5.0], [5.2]], [[0.0], [5.0]])
    >>> labels
    [0, 0, 1, 1]
    """
    Xa = _as_matrix(X)
    C = _as_centroids(centroids, Xa.shape[1])
    d2 = _pairwise_sq_dist(Xa, C)
    labels = np.argmin(d2, axis=1)
    new_c = C.copy()
    for j in range(C.shape[0]):
        members = Xa[labels == j]
        if members.shape[0] > 0:
            new_c[j] = members.mean(axis=0)
    return labels.tolist(), new_c.tolist()


def inertia(X: Matrix, centroids: Matrix) -> float:
    """Within-cluster sum of squared distances to the nearest centroid.

    This is the K-Means objective ``J(C) = sum_i min_j ||x_i - c_j||^2``.
    """
    Xa = _as_matrix(X)
    C = _as_centroids(centroids, Xa.shape[1])
    d2 = _pairwise_sq_dist(Xa, C)
    return float(d2.min(axis=1).sum())


def mini_batch_update(
    centroids: Matrix,
    counts: Sequence[float],
    batch: Matrix,
) -> tuple[list[list[float]], list[float]]:
    """One mini-batch K-Means update (Sculley, 2010).

    Each point in ``batch`` is assigned to its nearest current centroid; the per
    centroid count is incremented and the centroid is moved toward the point with
    learning rate ``eta = 1 / count``. Points are processed in order, so a centroid
    moves as later points in the same batch are absorbed (the running-mean form of
    SGD on the K-Means objective).

    Returns ``(new_centroids, new_counts)``.
    """
    C = np.asarray(centroids, dtype=float)
    if C.ndim != 2:
        raise ValueError("centroids must be a 2-D array of shape (k, n_features)")
    cnt = np.asarray(counts, dtype=float)
    if cnt.shape[0] != C.shape[0]:
        raise ValueError(f"counts length {cnt.shape[0]} does not match number of centroids {C.shape[0]}")
    if np.any(cnt < 0):
        raise ValueError("counts must be non-negative")
    B = _as_matrix(batch, name="batch")
    if B.shape[1] != C.shape[1]:
        raise ValueError(f"batch dimension {B.shape[1]} does not match centroid dimension {C.shape[1]}")
    C = C.copy()
    cnt = cnt.copy()
    for x in B:
        d2 = np.sum((C - x) ** 2, axis=1)
        j = int(np.argmin(d2))
        cnt[j] += 1.0
        eta = 1.0 / cnt[j]
        C[j] = (1.0 - eta) * C[j] + eta * x
    return C.tolist(), cnt.tolist()


def kmedians_centroid(members: Matrix) -> list[float]:
    """Coordinatewise median, the L1-optimal cluster representative.

    The minimizer of total L1 deviation along each axis is the per-dimension
    median, so this is the k-medians update step. For an even number of points the
    lower of the two central order statistics is returned, matching the convention
    used by the Julia and Rust implementations so the three agree exactly.
    """
    M = _as_matrix(members, name="members")
    n = M.shape[0]
    out = []
    for d in range(M.shape[1]):
        col = np.sort(M[:, d])
        # Lower median for even n keeps all three languages bit-identical.
        idx = (n - 1) // 2
        out.append(float(col[idx]))
    return out


def pam_assign_cost(distances: Matrix, medoid_indices: Sequence[int]) -> tuple[list[int], float]:
    """Assign every point to its nearest medoid given a dissimilarity matrix.

    ``distances`` is an ``n x n`` matrix of arbitrary (possibly non-metric)
    dissimilarities; ``medoid_indices`` selects ``k`` rows/columns as the current
    medoids. Returns ``(labels, total_cost)`` where ``labels[i]`` is the position
    within ``medoid_indices`` of the medoid nearest to point ``i`` and
    ``total_cost`` is the sum of those nearest dissimilarities (the k-medoids
    objective). This is the assignment core shared by PAM and FastPAM.
    """
    D = np.asarray(distances, dtype=float)
    if D.ndim != 2 or D.shape[0] != D.shape[1]:
        raise ValueError("distances must be a square n x n dissimilarity matrix")
    n = D.shape[0]
    med = list(medoid_indices)
    if not med:
        raise ValueError("medoid_indices must be non-empty")
    for m in med:
        if not 0 <= m < n:
            raise ValueError(f"medoid index {m} out of range [0, {n})")
    labels = []
    total = 0.0
    for i in range(n):
        best_pos, best_cost = 0, math.inf
        for pos, m in enumerate(med):
            c = D[i, m]
            if c < best_cost:
                best_cost, best_pos = c, pos
        labels.append(best_pos)
        total += best_cost
    return labels, float(total)


def rbf_kernel_matrix(X: Matrix, gamma: float) -> list[list[float]]:
    """Gaussian (RBF) kernel matrix ``K[i,j] = exp(-gamma ||x_i - x_j||^2)``."""
    if gamma <= 0:
        raise ValueError(f"gamma must be positive, got {gamma}")
    Xa = _as_matrix(X)
    d2 = _pairwise_sq_dist(Xa, Xa)
    return np.exp(-gamma * d2).tolist()


def kernel_assignment_distances(kernel: Matrix, labels: Sequence[int], n_clusters: int) -> list[list[float]]:
    """Feature-space squared distances from each point to every cluster mean.

    Uses the kernel trick: with a precomputed kernel ``K`` and a hard assignment,
    the squared distance from ``phi(x_i)`` to the mean of cluster ``j`` is

        K[i,i] - (2/|S_j|) sum_{l in S_j} K[i,l] + (1/|S_j|^2) sum_{l,m in S_j} K[l,m].

    Returns an ``n x n_clusters`` matrix of these distances; an empty cluster
    yields ``+inf`` for every point (so it is never chosen by argmin).
    """
    K = np.asarray(kernel, dtype=float)
    if K.ndim != 2 or K.shape[0] != K.shape[1]:
        raise ValueError("kernel must be a square n x n matrix")
    n = K.shape[0]
    lab = list(labels)
    if len(lab) != n:
        raise ValueError(f"labels length {len(lab)} does not match kernel size {n}")
    if n_clusters <= 0:
        raise ValueError("n_clusters must be positive")
    for c in lab:
        if not 0 <= c < n_clusters:
            raise ValueError(f"label {c} out of range [0, {n_clusters})")

    members = [[i for i in range(n) if lab[i] == j] for j in range(n_clusters)]
    third = []  # constant per-cluster term
    for S in members:
        if S:
            sub = K[np.ix_(S, S)]
            third.append(float(sub.sum()) / (len(S) ** 2))
        else:
            third.append(math.inf)

    out = []
    for i in range(n):
        row = []
        for j in range(n_clusters):
            S = members[j]
            if not S:
                row.append(math.inf)
                continue
            cross = float(K[i, S].sum())
            d = K[i, i] - 2.0 * cross / len(S) + third[j]
            row.append(d)
        out.append(row)
    return out


def fuzzy_centroids(X: Matrix, memberships: Matrix, m: float) -> list[list[float]]:
    """Membership-weighted cluster centers for fuzzy c-means.

    ``c_j = sum_i u_ij^m x_i / sum_i u_ij^m`` with fuzziness exponent ``m > 1``.
    """
    if m <= 1.0:
        raise ValueError(f"fuzziness exponent m must be greater than 1, got {m}")
    Xa = _as_matrix(X)
    U = np.asarray(memberships, dtype=float)
    if U.ndim != 2 or U.shape[0] != Xa.shape[0]:
        raise ValueError("memberships must have shape (n_samples, n_clusters)")
    w = U ** m
    denom = w.sum(axis=0)
    if np.any(denom == 0.0):
        raise ValueError("a cluster has zero total membership; cannot form its centroid")
    return ((w.T @ Xa) / denom[:, None]).tolist()


def fuzzy_memberships(X: Matrix, centroids: Matrix, m: float) -> list[list[float]]:
    """Update fuzzy memberships from distances to all centroids (Bezdek FCM).

    ``u_ij = 1 / sum_l (||x_i - c_j|| / ||x_i - c_l||)^{2/(m-1)}``. A point that
    coincides exactly with one or more centroids is assigned membership split
    uniformly across those zero-distance centroids and zero elsewhere.
    """
    if m <= 1.0:
        raise ValueError(f"fuzziness exponent m must be greater than 1, got {m}")
    Xa = _as_matrix(X)
    C = _as_centroids(centroids, Xa.shape[1])
    dist = np.sqrt(_pairwise_sq_dist(Xa, C))  # (n, k)
    p = 2.0 / (m - 1.0)
    n, k = dist.shape
    out = np.zeros((n, k), dtype=float)
    for i in range(n):
        zero = np.where(dist[i] == 0.0)[0]
        if zero.size > 0:
            out[i, zero] = 1.0 / zero.size
            continue
        ratios = (dist[i][:, None] / dist[i][None, :]) ** p  # (k, k)
        out[i] = 1.0 / ratios.sum(axis=1)
    return out.tolist()


def bisecting_split(X: Matrix, init_two_centroids: Matrix) -> tuple[list[int], list[list[float]], float]:
    """Run 2-means to convergence to bisect a cluster.

    Given the points of a cluster and two initial centroids, alternates Lloyd
    steps until the assignment stops changing, returning
    ``(labels, two_centroids, sse)`` where ``labels`` is in ``{0, 1}`` and ``sse``
    is the resulting within-cluster sum of squares. This is the inner split that
    bisecting K-Means repeats top-down to build a divisive hierarchy.
    """
    Xa = _as_matrix(X)
    C = _as_centroids(init_two_centroids, Xa.shape[1])
    if C.shape[0] != 2:
        raise ValueError(f"bisecting requires exactly 2 initial centroids, got {C.shape[0]}")
    prev = None
    centroids = C.tolist()
    labels: list[int] = []
    for _ in range(1000):
        labels, centroids = lloyd_step(Xa.tolist(), centroids)
        if prev is not None and labels == prev:
            break
        prev = labels
    return labels, centroids, inertia(Xa.tolist(), centroids)
